capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+
+// Resources for SD Card API
+// elm-chan.org/docs/mmc/mmc_e.html
+// alumni.cs.ucr.edu/~amitra/sdcard/Additional/sdcard_appnote_foust.pdf
+// luckyresistor.me/cat-protector/software/sdcard-2/
+// http://users.ece.utexas.edu/~valvano/EE345M/SD_Physical_Layer_Spec.pdf
+
+use core::cell::Cell;
+use core::cmp;
+use kernel::{AppId, AppSlice, Callback, Driver, ReturnCode, Shared};
+use kernel::common::take_cell::{MapCell, TakeCell};
+
+use kernel::hil;
+use kernel::hil::time::Frequency;
+
+// Buffers used for SD card transactions
+// Constraints:
+//  * RXBUFFER must be greater than or equal to TXBUFFER in length
+//  * Both RXBUFFER and TXBUFFER must be longer  than the SD card's block size
+pub static mut TXBUFFER: [u8; 515] = [0; 515];
+pub static mut RXBUFFER: [u8; 515] = [0; 515];
+
+// Buffer for SD card driver
+pub static mut KERNEL_BUFFER: [u8; 512] = [0; 512];
+
+#[allow(dead_code)]
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum SDCmd {
+    CMD0 = 0, // Reset
+    CMD1 = 1, // Generic init
+    CMD8 = 8, // Check voltage range
+    CMD9 = 9, // Read chip specific data (CSD) register
+    CMD12 = 12, // Stop multiple block read
+    CMD16 = 16, // Set blocksize
+    CMD17 = 17, // Read single block
+    CMD18 = 18, // Read multiple blocks
+    CMD24 = 24, // Write single block
+    CMD25 = 25, // Write multiple blocks
+    CMD55 = 55, // Next command will be application specific
+    CMD58 = 58, // Read operation condition register (OCR)
+    ACMD41 = 0x80 + 41, // App-specific Init
+}
+
+#[allow(dead_code)]
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum SDResponse {
+    R1,
+    R2,
+    R3,
+    R7,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum State {
+    Idle,
+
+    SendACmd { acmd: SDCmd, arg: u32 },
+
+    InitReset,
+    InitCheckVersion,
+    InitRepeatHCSInit,
+    InitCheckCapacity,
+    InitAppSpecificInit,
+    InitRepeatAppSpecificInit,
+    InitRepeatGenericInit,
+    InitSetBlocksize,
+    InitComplete,
+
+    StartReadBlocks { count: u32 },
+    WaitReadBlock,
+    ReadBlockComplete,
+    WaitReadBlocks { count: u32 },
+    ReceivedBlock { count: u32 },
+    ReadBlocksComplete,
+
+    StartWriteBlocks { count: u32 },
+    WriteBlockResponse,
+    WriteBlockBusy,
+    WaitWriteBlockBusy,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum AlarmState {
+    Idle,
+
+    DetectionChange,
+
+    RepeatHCSInit,
+    RepeatAppSpecificInit,
+    RepeatGenericInit,
+
+    WaitForDataBlock,
+    WaitForDataBlocks { count: u32 },
+
+    WaitForWriteBusy,
+}
+
+#[derive(Clone,Copy,Debug,PartialEq)]
+enum ErrorCode {
+    CardStateChanged = -1,
+    InitializationFailure = -2,
+    ReadFailure = -3,
+    WriteFailure = -4,
+    TimeoutFailure = -5,
+}
+
+// SD Card types
+const CT_MMC: u8 = 0x01;
+const CT_SD1: u8 = 0x02;
+const CT_SD2: u8 = 0x04;
+const CT_BLOCK: u8 = 0x08;
+
+// SD Card capsule, capable of being built on top of by other kernel capsules
+pub struct SDCard<'a, A: hil::time::Alarm + 'a> {
```

Comment:
This is a personal preference, but I like it when the main struct is first in a module, since it lays out, at a high level, what the context for everything else is.